simulator/generation/mod.rs

```diff
@@ -60,6 +67,36 @@ pub(crate) fn one_of<'a, T, R: Rng>(choices: Vec<Box<dyn Fn(&mut R) -> T + 'a>>,
     choices[index](rng)
 }
 
+/// backtrack is a helper function for composing different "failable" generators.
+/// The function takes a list of functions that return an Option<T>, along with number of retries
+/// to make before giving up.
+pub(crate) fn backtrack<'a, T, R: Rng>(
+    mut choices: Vec<(u32, Box<dyn Fn(&mut R) -> Option<T> + 'a>)>,
+    rng: &mut R,
+) -> T {
+    loop {
+        // If there are no more choices left, we give up
+        let choices_ = choices
+            .iter()
+            .enumerate()
+            .filter(|(_, (retries, _))| *retries > 0)
+            .collect::<Vec<_>>();
+        if choices_.is_empty() {
+            panic!("backtrack: no more choices left");
```

Comment:
wait why is this panicing?

Reply:
If backtrack is total, then we it needs to return `Option<T>` too. Which means its consumer would need to handle the failure. Up until the top(the level of each property), there's not a very good point to handle the failure, so we would need to discard all the generation. That's why I made the choice of making it a partial function. Another option would be to receive at least 1 total generator(`Fn(Rng) -> T`) that will be the last resort if all others fail.